File: ios-native/Sprout/Model/Workshop/Sample.swift
import Foundation

/// Разбор снимка для объёмной модели: годится ли он и какие на нём цвета.
/// Только арифметика над пикселями: сам снимок и маску растения даёт
/// `Eye.study`, а здесь всё проверяется без телефона.
enum Sample {
    enum Verdict: Equatable, Sendable {
        case fine
        /// Мыльный: смазан или не в фокусе.
        case blurry
        case dark
        case bright
        /// Растения на снимке не видно.
        case empty

        var line: String {
            switch self {
            case .fine: Lang.text("Модель построю по снимку.")
            case .blurry: Lang.text("Снимок мыльный — снимите почётче.")
            case .dark: Lang.text("Снимок тёмный — снимите при свете.")
            case .bright: Lang.text("Снимок пересвечен — снимите без прямого солнца.")
            case .empty: Lang.text("Растения на снимке не видно — снимите его целиком.")
            }
        }
    }

    struct Reading: Equatable, Sendable {
        var verdict: Verdict
        var traits: Traits? = nil
    }

    /// Резкость ниже этого — мыльный снимок. Порог мягкий: отсеять надо
    /// смазанное, а не просто не идеальное.
    static let sharpEnough: Float = 40

    /// Пиксели RGBA рядами сверху вниз; маска — байт на пиксель, растение
    /// ярче половины. Нет маски — растение ищется в середине кадра.
    static func read(rgba: [UInt8], mask: [UInt8]?, width: Int,
                     height: Int) -> Reading {
        guard width > 8, height > 8, rgba.count >= width * height * 4 else {
            return Reading(verdict: .empty)
        }
        let inside: [Bool]
        if let mask, mask.count >= width * height {
            inside = mask.map { $0 > 127 }
        } else {
            inside = centre(width, height)
        }
        var box = (left: width, right: -1, top: height, bottom: -1)
        var count = 0
        var light: Float = 0
        for y in 0 ..< height {
            for x in 0 ..< width where inside[y * width + x] {
                count += 1
                box.left = min(box.left, x)
                box.right = max(box.right, x)
                box.top = min(box.top, y)
                box.bottom = max(box.bottom, y)
                let at = (y * width + x) * 4
                light += (0.3 * Float(rgba[at]) + 0.59 * Float(rgba[at + 1])
                    + 0.11 * Float(rgba[at + 2])) / 255
            }
        }
        guard count > width * height / 50 else {
            return Reading(verdict: .empty)
        }
        let mean = light / Float(count)
        if mean < 0.08 { return Reading(verdict: .dark) }
        if mean > 0.93 { return Reading(verdict: .bright) }
        if sharpness(rgba, width: width, height: height, box: box)
            < sharpEnough {
            return Reading(verdict: .blurry)
        }

        var leaves: [[UInt8]] = [[], [], []]
        var pale: [[UInt8]] = [[], [], []]
        var petals: [Int: [[UInt8]]] = [:]
        var pot: [[UInt8]] = [[], [], []]
        let boxHeight = max(box.bottom - box.top, 1)
        for y in box.top ... box.bottom {
            // Нижняя треть — горшок; цветы ищутся только выше него.
            let low = Float(y - box.top) / Float(boxHeight) > 0.65
            for x in box.left ... box.right where inside[y * width + x] {
                let at = (y * width + x) * 4
                let (hue, saturation, value) = hsv(rgba[at], rgba[at + 1],
                                                    rgba[at + 2])
                let green = hue >= 55 && hue <= 175 && saturation >= 0.18
                    && value >= 0.12
                func keep(_ bucket: inout [[UInt8]]) {
                    bucket[0].append(rgba[at])
                    bucket[1].append(rgba[at + 1])
                    bucket[2].append(rgba[at + 2])
                }
                if green {
                    keep(&leaves)
                } else if low {
                    keep(&pot)
                } else if saturation < 0.2 && value > 0.72 {
                    keep(&pale)
                } else if saturation >= 0.35 && value >= 0.3 {
                    let bin = Int(hue / 30) % 12
                    var bucket = petals[bin] ?? [[], [], []]
                    keep(&bucket)
                    petals[bin] = bucket
                }
            }
        }
        guard leaves[0].count > count / 50 else {
            return Reading(verdict: .empty)
        }
        let leaf = median(leaves)
        let variegation = pale[0].count > leaves[0].count / 8
            ? median(pale) : nil
        let bloom = petals.max { $0.value[0].count < $1.value[0].count }
        let flower = (bloom?.value[0].count ?? 0) > count / 100
            ? bloom.map { median($0.value) } : nil
        let pottery = pot[0].count > count * 3 / 100 ? median(pot) : nil
        let area = Float((box.right - box.left + 1) * (box.bottom - box.top + 1))
        let thick = Float(leaves[0].count) / area
        let density = 0.75 + 0.6 * Double(min(max((thick - 0.15) / 0.45, 0), 1))
        let ratio = Double(box.bottom - box.top + 1)
            / Double(box.right - box.left + 1)
        let stretch = min(max(ratio / 1.15, 0.8), 1.3)
        return Reading(verdict: .fine, traits: Traits(
            leaf: leaf, variegation: variegation, flower: flower,
            pot: pottery, density: density, stretch: stretch))
    }

    /// Разброс лапласиана по яркости: у резкого снимка края рвутся, у
    /// мыльного плавные.
    static func sharpness(_ rgba: [UInt8], width: Int, height: Int,
                          box: (left: Int, right: Int, top: Int,
                                bottom: Int)) -> Float {
        func gray(_ x: Int, _ y: Int) -> Float {
            let at = (y * width + x) * 4
            return 0.3 * Float(rgba[at]) + 0.59 * Float(rgba[at + 1])
                + 0.11 * Float(rgba[at + 2])
        }
        let left = max(box.left, 1)
        let right = min(box.right, width - 2)
        let top = max(box.top, 1)
        let bottom = min(box.bottom, height - 2)
        guard left < right, top < bottom else { return 0 }
        var sum: Float = 0
        var square: Float = 0
        var count: Float = 0
        for y in top ... bottom {
            for x in left ... right {
                let edge = gray(x - 1, y) + gray(x + 1, y) + gray(x, y - 1)
                    + gray(x, y + 1) - 4 * gray(x, y)
                sum += edge
                square += edge * edge
                count += 1
            }
        }
        let mean = sum / count
        return square / count - mean * mean
    }

    /// Средний по порядку — не среднее: пара бликов цвет не сдвинет.
    static func median(_ channels: [[UInt8]]) -> Channels {
        func middle(_ values: [UInt8]) -> Double {
            let sorted = values.sorted()
            return Double(sorted[sorted.count / 2])
        }
        return Channels(middle(channels[0]), middle(channels[1]),
                        middle(channels[2]))
    }

    /// Тон в градусах, насыщенность и яркость 0…1.
    static func hsv(_ r: UInt8, _ g: UInt8, _ b: UInt8)
        -> (hue: Float, saturation: Float, value: Float) {
        let red = Float(r) / 255
        let green = Float(g) / 255
        let blue = Float(b) / 255
        let top = max(red, green, blue)
        let bottom = min(red, green, blue)
        let spread = top - bottom
        guard spread > 1e-5 else { return (0, 0, top) }
        var hue: Float
        if top == red {
            hue = (green - blue) / spread
        } else if top == green {
            hue = (blue - red) / spread + 2
        } else {
            hue = (red - green) / spread + 4
        }
        hue *= 60
        if hue < 0 { hue += 360 }
        return (hue, spread / top, top)
    }

    /// Без маски растение ищется в середине кадра — эллипс на семь десятых.
    private static func centre(_ width: Int, _ height: Int) -> [Bool] {
        var inside = [Bool](repeating: false, count: width * height)
        for y in 0 ..< height {
            for x in 0 ..< width {
                let dx = (Float(x) / Float(width) - 0.5) / 0.35
                let dy = (Float(y) / Float(height) - 0.5) / 0.35
                inside[y * width + x] = dx * dx + dy * dy <= 1
            }
        }
        return inside
    }
}

File: android/app/src/main/kotlin/com/ledro6/sprout/model/Sample.kt
package com.ledro6.sprout.model

import kotlin.math.max
import kotlin.math.min

/** Квадратный кадр, вырезанный из снимка, — в точках снимка, а не экрана. */
data class Crop(val x: Double, val y: Double, val side: Double) {
    fun inside(width: Double, height: Double): Boolean {
        val slip = 1e-6
        return x >= -slip && y >= -slip && x + side <= width + slip && y + side <= height + slip
    }

    companion object {
        /** Дальше снимок рассыпается на точки. */
        const val DEEPEST = 4.0

        /**
         * Кадр, видимый в окне: снимок вписан «враспор», увеличен в `scale` раз и
         * сдвинут на `offset`. Сдвиг держится в границах — пустого угла не будет.
         */
        fun of(width: Double, height: Double, window: Double, scale: Double, offsetX: Double, offsetY: Double): Crop {
            val w = max(width, 1.0)
            val h = max(height, 1.0)
            val side = max(window, 1.0)
            val cover = max(side / w, side / h)
            val zoom = cover * scale.coerceIn(1.0, DEEPEST)
            val (heldX, heldY) = hold(offsetX, offsetY, width, height, window, scale)
            val left = (w * zoom - side) / 2 - heldX
            val top = (h * zoom - side) / 2 - heldY
            return Crop(left / zoom, top / zoom, side / zoom)
        }

        fun hold(offsetX: Double, offsetY: Double, width: Double, height: Double, window: Double, scale: Double): Pair<Double, Double> {
            val (roomX, roomY) = slack(width, height, window, scale)
            return offsetX.coerceIn(-roomX, roomX) to offsetY.coerceIn(-roomY, roomY)
        }

        /** Сколько снимок больше окна с каждой стороны; ноль — двигать некуда. */
        fun slack(width: Double, height: Double, window: Double, scale: Double): Pair<Double, Double> {
            val w = max(width, 1.0)
            val h = max(height, 1.0)
            val side = max(window, 1.0)
            val zoom = max(side / w, side / h) * scale.coerceIn(1.0, DEEPEST)
            return max(0.0, (w * zoom - side) / 2) to max(0.0, (h * zoom - side) / 2)
        }
    }
}

/**
 * Разбор снимка для своей модели: годится ли он и какие на нём цвета. Только
 * арифметика над пикселями — проверяется без телефона.
 */
object Sample {
    enum class Verdict {
        FINE, BLURRY, DARK, BRIGHT, EMPTY;

        val line: String
            get() = when (this) {
                FINE -> Lang.text("Модель построю по снимку.")
                BLURRY -> Lang.text("Снимок мыльный — снимите почётче.")
                DARK -> Lang.text("Снимок тёмный — снимите при свете.")
                BRIGHT -> Lang.text("Снимок пересвечен — снимите без прямого солнца.")
                EMPTY -> Lang.text("Растения на снимке не видно — снимите его целиком.")
            }
    }

    data class Reading(val verdict: Verdict, val traits: Traits? = null)

    const val SHARP_ENOUGH = 40f

    private class Bucket {
        val r = ArrayList<Int>()
        val g = ArrayList<Int>()
        val b = ArrayList<Int>()
        val size get() = r.size

        fun keep(rgba: ByteArray, at: Int) {
            r.add(rgba[at].toInt() and 0xFF)
            g.add(rgba[at + 1].toInt() and 0xFF)
            b.add(rgba[at + 2].toInt() and 0xFF)
        }

        fun median(): Channels {
            fun middle(values: List<Int>) = values.sorted()[values.size / 2].toDouble()
            return Channels(middle(r), middle(g), middle(b))
        }
    }

    /** Пиксели RGBA рядами сверху вниз; маска — байт на пиксель. Нет маски — середина кадра. */
    fun read(rgba: ByteArray, mask: ByteArray?, width: Int, height: Int): Reading {
        if (width <= 8 || height <= 8 || rgba.size < width * height * 4) return Reading(Verdict.EMPTY)
        val inside: BooleanArray = if (mask != null && mask.size >= width * height) {
            BooleanArray(width * height) { (mask[it].toInt() and 0xFF) > 127 }
        } else {
            centre(width, height)
        }
        var left = width
        var right = -1
        var top = height
        var bottom = -1
        var count = 0
        var light = 0f
        for (y in 0 until height) {
            for (x in 0 until width) {
                if (!inside[y * width + x]) continue
                count += 1
                left = min(left, x)
                right = max(right, x)
                top = min(top, y)
                bottom = max(bottom, y)
                val at = (y * width + x) * 4
                light += (0.3f * u(rgba[at]) + 0.59f * u(rgba[at + 1]) + 0.11f * u(rgba[at + 2])) / 255f
            }
        }
        if (count <= width * height / 50) return Reading(Verdict.EMPTY)
        val mean = light / count
        if (mean < 0.08f) return Reading(Verdict.DARK)
        if (mean > 0.93f) return Reading(Verdict.BRIGHT)
        if (sharpness(rgba, width, height, left, right, top, bottom) < SHARP_ENOUGH) return Reading(Verdict.BLURRY)

        val leaves = Bucket()
        val pale = Bucket()
        val petals = HashMap<Int, Bucket>()
        val pot = Bucket()
        val boxHeight = max(bottom - top, 1)
        for (y in top..bottom) {
            val low = (y - top).toFloat() / boxHeight > 0.65f
            for (x in left..right) {
                if (!inside[y * width + x]) continue
                val at = (y * width + x) * 4
                val (hue, saturation, value) = hsv(u(rgba[at]), u(rgba[at + 1]), u(rgba[at + 2]))
                val green = hue in 55f..175f && saturation >= 0.18f && value >= 0.12f
                when {
                    green -> leaves.keep(rgba, at)
                    low -> pot.keep(rgba, at)
                    saturation < 0.2f && value > 0.72f -> pale.keep(rgba, at)
                    saturation >= 0.35f && value >= 0.3f -> petals.getOrPut((hue / 30).toInt() % 12) { Bucket() }.keep(rgba, at)
                }
            }
        }
        if (leaves.size <= count / 50) return Reading(Verdict.EMPTY)
        val leaf = leaves.median()
        val variegation = if (pale.size > leaves.size / 8) pale.median() else null
        val bloom = petals.values.maxByOrNull { it.size }
        val flower = if ((bloom?.size ?: 0) > count / 100) bloom?.median() else null
        val pottery = if (pot.size > count * 3 / 100) pot.median() else null
        val area = ((right - left + 1) * (bottom - top + 1)).toFloat()
        val thick = leaves.size / area
        val density = 0.75 + 0.6 * ((thick - 0.15f) / 0.45f).coerceIn(0f, 1f)
        val ratio = (bottom - top + 1).toDouble() / (right - left + 1)
        val stretch = (ratio / 1.15).coerceIn(0.8, 1.3)
        return Reading(Verdict.FINE, Traits(leaf, variegation, flower, pottery, density, stretch))
    }

    private fun u(byte: Byte): Float = (byte.toInt() and 0xFF).toFloat()

    /** Разброс лапласиана по яркости: у резкого снимка края рвутся. */
    fun sharpness(rgba: ByteArray, width: Int, height: Int, boxLeft: Int, boxRight: Int, boxTop: Int, boxBottom: Int): Float {
        fun gray(x: Int, y: Int): Float {
            val at = (y * width + x) * 4
            return 0.3f * u(rgba[at]) + 0.59f * u(rgba[at + 1]) + 0.11f * u(rgba[at + 2])
        }
        val left = max(boxLeft, 1)
        val right = min(boxRight, width - 2)
        val top = max(boxTop, 1)
        val bottom = min(boxBottom, height - 2)
        if (left >= right || top >= bottom) return 0f
        var sum = 0f
        var square = 0f
        var count = 0f
        for (y in top..bottom) {
            for (x in left..right) {
                val edge = gray(x - 1, y) + gray(x + 1, y) + gray(x, y - 1) + gray(x, y + 1) - 4 * gray(x, y)
                sum += edge
                square += edge * edge
                count += 1
            }
        }
        val mean = sum / count
        return square / count - mean * mean
    }

    /** Тон в градусах, насыщенность и яркость 0…1. */
    fun hsv(r: Float, g: Float, b: Float): Triple<Float, Float, Float> {
        val red = r / 255
        val green = g / 255
        val blue = b / 255
        val top = maxOf(red, green, blue)
        val bottom = minOf(red, green, blue)
        val spread = top - bottom
        if (spread <= 1e-5f) return Triple(0f, 0f, top)
        var hue = when (top) {
            red -> (green - blue) / spread
            green -> (blue - red) / spread + 2
            else -> (red - green) / spread + 4
        }
        hue *= 60
        if (hue < 0) hue += 360
        return Triple(hue, spread / top, top)
    }

    /** Без маски растение ищется в середине кадра — эллипс на семь десятых. */
    private fun centre(width: Int, height: Int): BooleanArray {
        val inside = BooleanArray(width * height)
        for (y in 0 until height) {
            for (x in 0 until width) {
                val dx = (x.toFloat() / width - 0.5f) / 0.35f
                val dy = (y.toFloat() / height - 0.5f) / 0.35f
                inside[y * width + x] = dx * dx + dy * dy <= 1
            }
        }
        return inside
    }
}
